src/library/scala/PartialFunction.scala

```diff
@@ -192,6 +216,22 @@ object PartialFunction {
     }
   }
 
+  /** Composite function produced by `PartialFunction#andThen` method
+    */
+  private class Combined[-A, B, +C] (pf: PartialFunction[A, B], k: PartialFunction[B, C]) extends PartialFunction[A, C] with Serializable {
+    def isDefinedAt(x: A) = {
+      val b: B = pf.applyOrElse(x, checkFallback[B])
```

Comment:
@NthPortal Maybe caching `pf.applyOrElse` into a lazy val class variable can achieve better performance?

Reply:
Also since this method calls the pf.applyOrElse, in case that pf is inpure (modifies external variable), this can mess up the external variable value.

Reply:
- For the best perfomance `applyOrElse` should be used
- caching can be done outside `isDefinedAt` if it really needed, caching inside can lead to unexpected memory usage and other undesired effects

Reply:
Memory and performance concerns aside, correct me if I am wrong, any side effect will be evaluated twice, which will result in undesired behavior.

Reply:
If You need the results of combined function, call `applyOrElse`, both partial functions will be called at most once. It is impossible to check if `k` is defined on `pf`'s result, so if `pf` is side-effectful You should avoid calling `isDefinedAt`

Reply:
Besides, `PartialFunction` provides no contract of non-evaluating body while checking if it is definedAt some argument.

Reply:
In that case it's probably not a good idea to do this. This will break existing code like 

```
var systemSwitch=false
/*...*/
val f= ({case _ => systemSwitch = !systemSwitch}: PartialFunction[Any, Any]).andThen(g)
if(f.isDefinedAt(x)){
  f(x)
}
```
, for many applications could have side effect defined in the partial functions and calling apply inside isDefinedAt not only affects performance, but will result external status to alter unexpectedly. I expect a bug report to come after it's released.
@SethTisue

Reply:
@danslapman Bruce's argument seems persuasive to me, wdyt?

Reply:
Don't get me wrong - I am all for the purpose of this PR. We need `andThen` to take PartialFunction. But we probably should come up with a better solution without multiple execution of the same code.

Reply:
@SethTisue I think we can name this function some other way to avoid such cases with mutable functions. How about `andThenNarrow`?

Reply:
Or, maybe, `andThenPartial`

Reply:
Sorry I'm a bit late to this discussion.

- it seems to me that you couldn't use a `lazy val`, unless you expect the combined PF to only be used once
- caching the result of the first evaluation would prevent double evaluation in most cases (except for concurrent use or really strange synchronous uses)
- it's only one additional field - I'm not personally super concerned about memory usage

I'm actually more concerned about the inconsistent behaviour of the overloaded methods than the double evaluation of side-effecting PFs. Consider the following:

```scala
val f: PartialFunction[Int, Int] = { case x if x % 2 == 0 => x + 2 }
val g: PartialFunction[Int, Int] = { case x if x % 2 == 1 => x - 2 }

val c1 = f andThen g
val c2 = f andThen (g: Int => Int)
```

In the above snippet, `c1` and `c2` confusingly do different things, despite being the result of calling `andThen` on `f` with exactly the same object. For example, `c1.applyOrElse(2, 8)` throws an exception, but `c2.applyOrElse(2, 8)` returns `8`. For that reason, I think it is important to give this behaviour a different name.

(Of course, this problem would not exist if `Function1` extended `PartialFunction`, like it ought to. Unfortunately, that ship has not only sailed, but reached its destination, and then made a return trip.)

Reply:
>In the above snippet, c1 and c2 confusingly do different things

This PR makes `andThen` consistent in that way, that's why I think we should name this method `andThen` and nothing else.

> caching the result of the first evaluation would prevent double evaluation in most cases
> it's only one additional field - I'm not personally super concerned about memory usage

- in fact it would lead to memoizing results into a `mutable.Map`, which can lead to significant memory consumption and, even, to OOM 
- any cashing assumes that function is pure
- `applyOrElse` is existing and _documented_ way to call `PartialFunction` effectively, and as I said earlier, PartialFunction provides no contract of non-evaluating body while checking if it is defined at some argument. I think there is not excuses to keep current inconsistent and confusing behaviour.

Reply:
> This PR makes andThen consistent in that way

I'm not understanding how this PR makes it consistent - my example is what would happen according to this PR. Unless the implementation of `andThen(Function1)` is changed to check for `PartialFunction`, the behavior depends on the *static* type of a function, rather than its actual type. At least to me, this is confusing and inconsistent.

Reply:
>what would happen according to this PR

That's wrong, `c1.applyOrElse(2, 8)` currently throws an exception, but will not would not with new `andThen` implementation

Reply:
While caching would avoid executing the side effect twice, we're still in a weird situation where `isDefinedAt` of a combined PF executes parts of its side-effect. There's no way around this, `isDefinedAt` is part of the interface and needs to be implemented.

Caching the value can also be surprising because a call to `apply` (after `isDefinedAt` has been called somewhere) has no side-effect.

We have to pick one or the other evil, so it's a documentation issue. I think we should go ahead with the current design. The documentation should say that the resulting PF may apply the first PF and execute its side effect. (We could do an early return by checking `pf.isDefinedAt` first, and avoid evaluating `pf` in this case).

Then I think this PR is good to go.

Reply:
@lrytz added notes on behavior of `Combined`.